drivers/can/can_mspm0_canfd.c

```diff
@@ -0,0 +1,276 @@
+#ifdef CONFIG_CAN_FD_MODE
+	.set_timing_data = can_mcan_set_timing_data,
+	.timing_data_min = CAN_MCAN_TIMING_DATA_MIN_INITIALIZER,
+	.timing_data_max = CAN_MCAN_TIMING_DATA_MAX_INITIALIZER,
+#endif /* CONFIG_CAN_FD_MODE */
+};
+
+static const struct can_mcan_ops can_mspm0_canfd_ops = {
+	.read_reg = can_mspm0_canfd_read_reg,
+	.write_reg = can_mspm0_canfd_write_reg,
+	.read_mram = can_mspm0_canfd_read_mram,
+	.write_mram = can_mspm0_canfd_write_mram,
+	.clear_mram = can_mspm0_canfd_clear_mram,
+};
+
+#define can_mspm0_canfd_IRQ_CFG_FUNCTION(index)							\
+	static void config_can_##index(void)							\
+	{											\
+		IRQ_CONNECT(DT_INST_IRQN(index), DT_INST_IRQ(index, priority),			\
+			    can_mspm0_canfd_isr, DEVICE_DT_INST_GET(index), 0);			\
+		irq_enable(DT_INST_IRQN(index));						\
+	}
+
+#define can_mspm0_canfd_CFG_INST(index)								\
+	BUILD_ASSERT(CAN_MCAN_DT_INST_MRAM_ELEMENTS_SIZE(index) <=				\
+			     CAN_MCAN_DT_INST_MRAM_SIZE(index),					\
+		     "Insufficient Message RAM size");						\
+												\
+	PINCTRL_DT_INST_DEFINE(index);								\
+	CAN_MCAN_CALLBACKS_DEFINE(can_mspm0_canfd_cbs_##index,					\
+				  CAN_MCAN_DT_INST_MRAM_TX_BUFFER_ELEMENTS(index),		\
+				  CONFIG_CAN_MAX_STD_ID_FILTER, CONFIG_CAN_MAX_EXT_ID_FILTER);	\
+												\
+	static const struct mspm0_sys_clock mspm0_can_sys_clock##index =			\
+		MSPM0_CLOCK_SUBSYS_FN(index);							\
+												\
+	static const struct can_mspm0_canfd_config can_mspm0_canfd_cfg_##index = {		\
+		.ti_canfd_base = DT_REG_ADDR_BY_NAME(DT_DRV_INST(index), ti_canfd),		\
+		.clock_subsys = &mspm0_can_sys_clock##index,					\
+		.clock_cfg =									\
+			{									\
+				.clockSel = DT_PROP_OR(DT_DRV_INST(index),			\
+					ti_canclk_source, CAN_MSPM0_CLK_SRC_PLLCLK1) ==		\
+					CAN_MSPM0_CLK_SRC_PLLCLK1 ? DL_MCAN_FCLK_SYSPLLCLK1	\
+									: DL_MCAN_FCLK_HFCLK,	\
+				.divider = DT_PROP(DT_DRV_INST(index), ti_divider),		\
+			},									\
+		.mcan_base = CAN_MCAN_DT_INST_MCAN_ADDR(index),					\
+		.mram = CAN_MCAN_DT_INST_MRAM_ADDR(index),					\
+		.irq_cfg_func = config_can_##index,						\
+		.pinctrl = PINCTRL_DT_INST_DEV_CONFIG_GET(index),				\
+	};											\
+												\
+	static const struct can_mcan_config can_mcan_cfg_##index = CAN_MCAN_DT_CONFIG_INST_GET(	\
+		index, &can_mspm0_canfd_cfg_##index, &can_mspm0_canfd_ops,			\
+		&can_mspm0_canfd_cbs_##index);
+
+#define can_mspm0_canfd_DATA_INST(index)							\
+	static struct can_mcan_data can_mcan_data_##index = CAN_MCAN_DATA_INITIALIZER(NULL);
+
+#define can_mspm0_canfd_DEVICE_INST(index)							\
+	CAN_DEVICE_DT_INST_DEFINE(index, can_mspm0_canfd_init, NULL, &can_mcan_data_##index,	\
+				  &can_mcan_cfg_##index, POST_KERNEL, CONFIG_CAN_INIT_PRIORITY,	\
+				  &can_mspm0_canfd_driver_api);
+
+#define can_mspm0_canfd_INST(index)								\
+	can_mspm0_canfd_IRQ_CFG_FUNCTION(index) can_mspm0_canfd_CFG_INST(index)			\
+		can_mspm0_canfd_DATA_INST(index) can_mspm0_canfd_DEVICE_INST(index)
+
```

Comment:
I have not seen this construct before (especially wondering what's with the lower_UPPER case?). Please follow the naming used in other drivers and just inline these in one macro.